code/parachain/frame/pablo/src/testing.rs

```diff
@@ -0,0 +1,299 @@
+
+		let asset_1_id = PICA::ID;
+		let asset_2_id = USDT::ID;
+
+		Runtime::assert_extrinsic_event_with(
+			Pallet::<Runtime>::do_create_pool(
+				PoolInitConfiguration::DualAssetConstantProduct {
+					owner: ALICE.into(),
+					assets_weights: bounded_btree_map! {
+						asset_1_id.into() => Permill::from_parts(500_000),
+						asset_2_id.into() => Permill::from_parts(500_000),
+					},
+					fee: Permill::from_parts(10_000),
+				},
+				Some(PICA_USDT_LPT.into()),
+			),
+			|event| match event {
+				Event::PoolCreated { pool_id, .. } => Some(pool_id),
+				_ => None,
+			},
+		)
+	}
+
+	pub fn zero_fees_pool_1_4<Runtime: PabloRuntimeConstraints>() {
+		next_block::<Pallet<Runtime>, Runtime>();
+
+		let asset_1_id = PICA::ID;
+		let asset_2_id = USDT::ID;
+
+		let _pool_id = Runtime::assert_extrinsic_event_with(
+			Pallet::<Runtime>::do_create_pool(
+				PoolInitConfiguration::DualAssetConstantProduct {
+					owner: ALICE.into(),
+					assets_weights: bounded_btree_map! {
+						asset_1_id.into() => Permill::from_parts(500_000),
+						asset_2_id.into() => Permill::from_parts(500_000),
+					},
+					fee: Permill::zero(),
+				},
+				Some(PICA_USDT_LPT.into()),
+			),
+			|event| match event {
+				Event::PoolCreated { pool_id, .. } => Some(pool_id),
+				_ => None,
+			},
+		);
+	}
+
+	pub fn pool_assets_cannot_be_the_same_1_6<Runtime: PabloRuntimeConstraints>() {
+		next_block::<Pallet<Runtime>, Runtime>();
+
+		let asset_1_id = PICA::ID;
+		let asset_2_id = PICA::ID;
+
+		assert_noop!(
+			// don't need to use do_create_pool here as we're checking pool creation validations
+			// and never adding liquidity
+			Pallet::<Runtime>::create(
+				OriginFor::<Runtime>::root(),
+				PoolInitConfiguration::DualAssetConstantProduct {
+					owner: ALICE.into(),
+					assets_weights: bounded_btree_map! {
+						asset_1_id.into() => Permill::from_parts(500_000),
+						asset_2_id.into() => Permill::from_parts(500_000),
+					},
+					fee: Permill::zero(),
+				},
+			),
+			crate::Error::<Runtime>::InvalidPair,
+		);
+	}
+}
+
+pub mod providing_liquidity {
+	use super::*;
+
+	pub fn add_liquidity_to_1_1_pool_2_1<Runtime: PabloRuntimeConstraints>() {
+		next_block::<Pallet<Runtime>, Runtime>();
+
+		let pool_id = super::pool_creation::create_new_constant_product_pool_1_1::<Runtime>();
+
+		mint_assets::<Runtime>(PICA::ID, BOB, PICA::units(1_100_000));
+		mint_assets::<Runtime>(USDT::ID, BOB, USDT::units(1_100_000));
+
+		let assets = [
+			(PICA::ID.into(), PICA::units(10_000).into()),
+			(USDT::ID.into(), USDT::units(10_000).into()),
+		]
+		.into_iter()
+		.collect::<BTreeMap<_, _>>();
+
+		Runtime::assert_extrinsic_event(
+			Pallet::<Runtime>::add_liquidity(
+				OriginFor::<Runtime>::signed(BOB.into()),
+				pool_id,
+				assets.clone(),
+				Zero::zero(),
+				true,
+			),
+			Event::<Runtime>::LiquidityAdded {
+				who: BOB.into(),
+				pool_id,
+				asset_amounts: assets,
+				minted_lp: 19_999_999_993_470_955_u128.into(),
```

Comment:
Lets depend on the functions from `composable-math` here instead of a hard-coded value. If we want more testing for the math functions correctness, we can add those to the `composable-math` crate. This also applies to similar numbers throughout this test.

Reply:
good point, updated